Score observer pattern

FT-Fumetsu opened the pull request:
![Diagramme UML Observer Pattern](https://github.com/user-attachments/assets/275d72cd-36bd-46c2-9ea3-023cbb900a4e)

FT-Fumetsu committed:
main Observer Pattern scripts added

FT-Fumetsu committed:
Add TextMeshPro and GameObjects

J'ai ajouté TextMeshPro et le gameobject score, les ui et le gameobject BrickDestroyNotifier pour l'observer pattern

FT-Fumetsu committed:
Add Score Data

FT-Fumetsu committed:
Implement Score

LahessAimere commented on Assets/05-Scripts/Score/Score.cs:
```diff
@@ -0,0 +1,25 @@
+
+    public void OnNotify()
+    {
+        //Debug.Log("Score is NOTIFIED");
```
Si tu te sers plus du `debug` tu peux le retirer

LahessAimere commented on Assets/05-Scripts/Score/Score.cs:
```diff
@@ -0,0 +1,25 @@
+using UnityEngine;
+using ScoreObserverPattern;
+
+public class Score : MonoBehaviour, IObserver
```
Change peut être le nom en `ScoreObserver` pour que l'on sache direct ce que c'est

LahessAimere commented on Assets/05-Scripts/Score/Score.cs:
```diff
@@ -0,0 +1,25 @@
+
+public class Score : MonoBehaviour, IObserver
+{
+    [SerializeField] private ObserverSubject _playerSubject;
```
Ah j'avais pas compris ça pour moi c'était les brick, enfin pour moi tu dois être abonné à BrickDestroyNotifier, pas au joueur
Parce que quand la brick est cassé ça te donne des points, dis moi ce que tu en penses

FT-Fumetsu replied:
ah oui oui je me suis trompé, j'ai du avoir un moment d'absence xD

LahessAimere reviewed (changes requested):
Dans l'ensemble c'est bien juste quelques trucs à revoir

LahessAimere commented:
Ah et j'ai oublié de te dire mais quand t'as fini cette branch peut être rajoute dans le read me ce que t'as rajouté et comment ça fonctionne

FT-Fumetsu committed:
Update with PR

LahessAimere commented on Assets/05-Scripts/Score/ScoreObserver.cs:
```diff
@@ -1,25 +1,24 @@
     private void OnDisable()
     {
-        _playerSubject.RemoveObserver(this);
+        _observerSubject.RemoveObserver(this);
```
Du coup ça fonctionne plus

Utilises ton singleton `BrickDestroyNotifier` et écoute directement son instance

LahessAimere reviewed (changes requested):
Ah oui et très très important rajoute des namespaces à tous tes scripts

FT-Fumetsu committed:
Add Namespaces

FT-Fumetsu committed:
Update with BrickDestroyNotifier.Instance

FT-Fumetsu committed:
remove useless variable

FT-Fumetsu commented:
[Why Observer Pattern.pdf](https://github.com/user-attachments/files/20480853/Why.Observer.Pattern.pdf)

LahessAimere commented on Assets/05-Scripts/Score/Observer/BrickDestroyNotifier.cs:
```diff
@@ -0,0 +1,18 @@
+namespace ScoreObserverPattern
```
Tu as mis le même namespace partout
regardes dans mes scripts comment c'est décomposé

LahessAimere replied:
par exemple namespace : BrickBreaker.Brick

LahessAimere commented on Assets/05-Scripts/Score/Observer/IObserver.cs:
```diff
@@ -0,0 +1,7 @@
+namespace ScoreObserverPattern
```
Autre exemple : BrickBreaker.Score.Observer

LahessAimere commented on Assets/05-Scripts/bricks-generation/Types/StandardBrick.cs:
```diff
@@ -1,11 +1,18 @@
     {
         [SerializeField] protected int _hitPoints = 1;
 
+        //Test de l'observer sans d�truire les bricks parce que la balle n'est pas encore impl�ment�e
+        private void Start()
+        {
+            BrickDestroyNotifier.Instance?.NotifyBrickDestroyed(); // Notify score
+        }
+
```
C'est censé faire quoi ici ?

FT-Fumetsu replied:
J'ai mis le commentaire au dessus, vu qu'on ne peux pas encore détruire les briques, c'est pour vérifier que la notification s'envoie bien. Quand on aura la balle, on pourra modifier cette partie là (l''enlever)

LahessAimere commented on Assets/05-Scripts/Score/ScoreUI.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        [SerializeField] private TextMeshProUGUI _scoreText;
+
+        public void UpdateScoreUI(int score)
```
Tu n'utilises jamais cette fonction

FT-Fumetsu replied:
Elle est appelée dans la hiérarchie dans ScoreSystem

LahessAimere replied:
![image](https://github.com/user-attachments/assets/526557d6-e4c2-4bef-971f-746d03bf1a29)

Moi ça me dit 0 référence

LahessAimere replied:
![image](https://github.com/user-attachments/assets/9e14a97c-a0db-479b-a836-7daec952fa1f)

aaah j'avais pas compris d'accord

LahessAimere replied:
https://discord.com/channels/1022084643069112350/1040239907685290045/1358791622053331094

regardes là discussion pour t'aider pour faire le score

LahessAimere replied:
et si tu veux faire un event utilise plutôt Action<T> qu'un unity event
Pour le système de score, il vaut mieux utiliser un event en C# pour que ce soit propre, stable et facile à maintenir. Le UnityEvent, c’est pratique pour lier des choses dans l’inspecteur, mais ce n’est pas l’idéal pour de la logique comme le score.

UnityEvent c'est :
- Moins performant (très léger, mais vrai).
- Moins sûr au runtime : si une référence dans l’inspecteur est manquante ou mal liée, Unity ne crashe pas, mais rien ne se passe.
- Moins flexible pour la logique complexe, ou les callbacks multiples.

LahessAimere reviewed (changes requested):
### Namespace
_Changes tous tes namespace, tu dois normalement avoir un truc qui ressemble avec ce genre de namespace :_

- BrickBreaker.Score.Observers
- BrickBreaker.Brick
- BrickBreaker.Score
- BrickBreaker.Score.Data
- BrickBreaker.Score.UI

### Missing script
_N'oublie pas de les enlever :_

![image](https://github.com/user-attachments/assets/7b8de8b3-b87f-4d75-bda9-de613e35a129)

FT-Fumetsu committed:
Update PR

Changement namespaces, delete les missing scripts

FT-Fumetsu committed:
Modifications PR

j'ai laissé en commentaire ```BrickDestroyNotifier.Instance``` pour laisser tester et voir, je l'enlèverais avant de merge la pr

LahessAimere commented on Assets/05-Scripts/bricks-generation/Types/BonusBrick.cs:
```diff
@@ -6,6 +7,12 @@ public class BonusBrick : StandardBrick
     {
         [SerializeField] private GameObject _bonusPrefab;
 
+        //Test de l'observer sans d�truire les bricks parce que la balle n'est pas encore impl�ment�e
+        private void Start()
+        {
+            BrickDestroyNotifier.Instance?.NotifyBrickDestroyed(); // Notify score
+        }
+
```
Il faudra pas oublier de l'enlevé après mais je le ferai quand je ferai les test avec la balle (j'écris ici pour me le rappeler)

LahessAimere reviewed (changes requested):
J'ai enfin la notif que tu m'avais parlé !

![image](https://github.com/user-attachments/assets/57e9873c-f684-4862-96f4-e5f214aac184)

FT-Fumetsu committed:
ModifyScore

Pour ajouter du score pour le moment, va sur le gameObject BrickDestroyNotifier et il y a un ContextMenu qui envoie la notif (qui s'appele add score). J'ai aussi rajouté une fonction ```RemoveScore()``` pour quand le joueur pert une vie, ça me parraît cool

LahessAimere commented on Assets/05-Scripts/Score/ScoreData.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public int Score => _score;
+
+        private void OnEnable()
+        {
+            _score = 0;
+        }
+
+        public void SetScore(int value)
+        {
+            _score = value;
+        }
+    }
+}
```
pas de fonction dans un scriptable il faut qu'il comporte que des données

LahessAimere reviewed (changes requested):
C'est presque ça on y est presque, juste un truc a changé

FT-Fumetsu committed:
Update ScoreData

LahessAimere commented on Assets/05-Scripts/Score/ScoreObserver.cs:
```diff
@@ -0,0 +1,29 @@
+            _scoreSystem.AddScore(_scorePerBrick);
+        }
+
+        private void OnEnable()
+        {
+            _brickDestroyNotifier.AddObserver(this);
+            //BrickDestroyNotifier.Instance.AddObserver(this);
+        }
+
+        private void OnDisable()
+        {
+            _brickDestroyNotifier.RemoveObserver(this);
+            //BrickDestroyNotifier.Instance.RemoveObserver(this);
+        }
+    }
```
Enlève les commentaires

LahessAimere reviewed (changes requested):
Ok c'est good juste les commentaires supprime les et j'approve la pr

FT-Fumetsu committed:
Final Score Push
